### PR TITLE
fix(Section): defaultIsOpen should behave as expected

### LINKED ISSUE #6322: Section defaultIsOpen does the opposite of the provided boolean
#### Environment

- __Package version(s)__: 5.2.0
- __Operating System__: Windows 11
- __Browser name and version__: Chrome 115

#### Code Sandbox

Link to a minimal repro (fork [this code sandbox](https://codesandbox.io/p/sandbox/blueprint-sandbox-2023-fjo3z4)): <https://codesandbox.io/p/sandbox/objective-easley-stwn3x>

#### Steps to reproduce

1. See linked code sandbox.

#### Actual behavior

The `Section` component `collapseProps.defaultIsOpen` does the opposite of what is expected, i.e. when `defaultIsOpen={true}` the collapse is _closed_, and when `defaultIsOpen={false}` the collapse is _open_.

#### Expected behavior

When `defaultIsOpen={true}` the collapse should initially be _open_.

#### Possible solution

The component defines an `useState` variable called `isCollapsed` which is initialized with the `defaultIsOpen` value. This value should be negated when initially provided to `useState`.

```ts
const [isCollapsed, setIsCollapsed] = React.useState<boolean>(!collapseProps?.defaultIsOpen ?? false);
```
<https://github.com/palantir/blueprint/blob/develop/packages/core/src/components/section/section.tsx#L112C67-L112C68>

### PR DESCRIPTION
**Pull request by nerdstep**
Fixes #6322

#### Checklist

- [x] Includes tests
- [x] Update documentation

#### Changes proposed in this pull request:

Negate `defaultIsOpen` value provided to `useState` so that it matches the expected `isCollapsed` behavior.

#### Reviewers should focus on:

Whether this fixes the desired behavior.

#### Screenshot

when `defaultIsOpen={true}`

image

when `defaultIsOpen={false}`

image

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/blueprint](https://github.com/palantir/blueprint), @nerdstep! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: packages/core/src/components/section/section.tsx [hunk 1]
```diff
@@ -109,7 +112,7 @@ export const Section: React.FC<SectionProps> = React.forwardRef((props, ref) =>
         title,
         ...htmlProps
     } = props;
-    const [isCollapsed, setIsCollapsed] = React.useState<boolean>(collapseProps?.defaultIsOpen ?? false);
+    const [isCollapsed, setIsCollapsed] = React.useState<boolean>(!collapseProps?.defaultIsOpen ?? false);
```
Comment (gluxon) [inline packages/core/src/components/section/section.tsx]: The `!` turns `collapseProps?.defaultIsOpen` into a boolean. Is `?? false` needed if the left side will never be `undefined` or `null`?

It is a bit verbose, but I think `collapseProps?.defaultIsOpen ?? true` conveys the new `@default true` pretty well. Perhaps we should invert after performing that defaulting logic?

```suggestion
    const [isCollapsed, setIsCollapsed] = React.useState<boolean>(!(collapseProps?.defaultIsOpen ?? true));
```

**Review by gluxon — COMMENTED** [review]
Thanks for the help fixing this!

File: packages/core/src/components/section/section.tsx (hunk 1 above)
Comment (nerdstep) [reply packages/core/src/components/section/section.tsx]: Agreed, that does make for more readable code! Updated the PR with the suggested change, including a comment to better explain the behavior. Also added a new test to assert the `defaultIsOpen={undefined}` case.

Comment (gluxon) [reply packages/core/src/components/section/section.tsx]: Thanks!

File: packages/docs-app/src/examples/core-examples/sectionExample.tsx [hunk 2]
```diff
@@ -134,8 +144,14 @@ export class SectionExample extends React.PureComponent<ExampleProps, SectionExa
         return (
             <Example options={options} {...this.props}>
                 <Section
+                    // A `key` is provided here to force the component to
+                    // re-mount when `defaultIsOpen` is changed, otherwise
+                    // the local state in the `Collapse` component is not
+                    // updated.
+                    key={String(defaultIsOpen)}
```
Comment (gluxon) [inline packages/docs-app/src/examples/core-examples/sectionExample.tsx:151]: Using `key` to reset state and animations is one of my favorite tricks in the pocket too. 🙂

**Review by pgoldberg — APPROVED** [review]
Thanks!

File: packages/docs-app/src/examples/core-examples/sectionExample.tsx (hunk 2 above)
Comment (nerdstep) [reply packages/docs-app/src/examples/core-examples/sectionExample.tsx:151]: Haha yeah it's a neat little trick! :)

**Pull request merged**